nav2_collision_monitor/src/polygon_base.cpp

```diff
@@ -0,0 +1,214 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/polygon_base.hpp"
+#include "nav2_collision_monitor/kinematics.hpp"
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+PolygonBase::PolygonBase(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string & polygon_name,
+  const std::string & base_frame_id,
+  const double simulation_time_step)
+: node_(node), polygon_name_(polygon_name), base_frame_id_(base_frame_id),
+  action_type_(DO_NOTHING), max_points_(-1), slowdown_ratio_(0.0), time_before_collision_(-1.0),
+  simulation_time_step_(simulation_time_step), visualize_(false)
+{
+  RCLCPP_INFO(logger_, "[%s]: Creating PolygonBase", polygon_name_.c_str());
+}
+
+PolygonBase::~PolygonBase()
+{
+  RCLCPP_INFO(logger_, "[%s]: Destroying PolygonBase", polygon_name_.c_str());
+}
+
+
+bool PolygonBase::configure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string polygon_topic;
+
+  if (!getParameters(polygon_topic)) {
+    return false;
+  }
+
+  if (visualize_) {
+    // Fill polygon points for future usage
+    std::vector<Point> poly;
+    getPolygon(poly);
+    for (const Point & p : poly) {
+      geometry_msgs::msg::Point32 p_s;
+      p_s.x = p.x;
+      p_s.y = p.y;
+      // p_s.z will remain 0.0
+      polygon_.points.push_back(p_s);
+    }
+
+    rclcpp::QoS polygon_qos = rclcpp::SystemDefaultsQoS();  // set to default
+    polygon_pub_ = node->create_publisher<geometry_msgs::msg::PolygonStamped>(
+      polygon_topic, polygon_qos);
+  }
+
+  return true;
+}
+
+void PolygonBase::activate()
+{
+  if (visualize_) {
+    polygon_pub_->on_activate();
+  }
+}
+
+void PolygonBase::deactivate()
+{
+  if (visualize_) {
+    polygon_pub_->on_deactivate();
+  }
+}
+
+ActionType PolygonBase::getActionType()
+{
+  return action_type_;
+}
+
+int PolygonBase::getMaxPoints()
+{
+  return max_points_;
+}
+
+double PolygonBase::getSlowdownRatio()
+{
+  return slowdown_ratio_;
+}
+
+double PolygonBase::getTimeBeforeCollision()
+{
+  return time_before_collision_;
+}
+
+double PolygonBase::getCollisionTime(
+  const std::vector<Point> & points, const Velocity & velocity)
+{
+  // Initial robot pose is {0,0} in base_footprint coordinates
+  Pose pose = {0, 0, 0};
+  Velocity vel = velocity;
+
+  // Robot movement simulation
+  for (double time = 0.0; time <= time_before_collision_; time += simulation_time_step_) {
+    // NOTE: vel is changing during the simulation
+    projectState(simulation_time_step_, pose, vel);
+    std::vector<Point> points_transformed = points;
+    transformPoints(pose, points_transformed);
+    if (getPointsInside(points_transformed) > 0) {
```

Comment:
should this always be `> 0`?

Reply:
Not, if there are no points this routine will return zero-value. This check confirms, that we found at least one point in a frame concerned with projected `pose`.

I thought to make one more function called `bool isPointsInside(points)`, that will be optimized for this case, skipping the inner cycle if there at least one collision found. However, I believe this overhead is not huge due to much more we spend on outer simulation cycle in `PolygonBase::getCollisionTime()`

Reply:
What I meant was should this be using `max_points` instead for some leniency for noise? Maybe you're right for on-approach we always trigger if there's any. 

Also, I noticed on second glance that `std::vector<Point> points_transformed = points;` is in the loop. If you preallocate `std::vector<Point> points_transformed;` then only set `points_transformed = points;` in the loop, then that will speed things up a bit when the vector is large

Reply:
[Fixed](https://github.com/ros-planning/navigation2/pull/2982/files#diff-e1c76d88179c14a1cf399e94bebd4f5d17dedca39e09b3bbdc810aa11a705543R564). Now it was moved to `collision_monitor.cpp`.

For the second notice about vector initialization, I've [moved](https://github.com/ros-planning/navigation2/pull/2982/files#diff-e1c76d88179c14a1cf399e94bebd4f5d17dedca39e09b3bbdc810aa11a705543R552) out of cycle vector declaration. Initialization of vector outside the loop should not give an effect, since it will be allocated inside the loop at first time, and then each next time when we are assigning a new value, it will be just copied but not reallocated.